pkg/storage/replica_command.go

```diff
@@ -2009,15 +2010,118 @@ func (s *Store) AdminRelocateRange(
 		return err
 	}
 	rangeDesc = *newDesc
-	rangeReplicas := rangeDesc.Replicas().All()
-	if len(rangeReplicas) != len(rangeDesc.Replicas().Voters()) {
-		// We just removed all the learners, so there shouldn't be anything but
-		// voters.
-		return crdberrors.AssertionFailedf(
-			`range %s had non-voter replicas: %v`, &rangeDesc, rangeDesc.Replicas())
+
+	canRetry := func(err error) bool {
+		whitelist := []string{
+			snapshotApplySemBusyMsg,
+			IntersectingSnapshotMsg,
+		}
+		for _, substr := range whitelist {
+			if strings.Contains(err.Error(), substr) {
+				return true
+			}
+		}
+		return false
+	}
+
+	startKey := rangeDesc.StartKey.AsRawKey()
+	transferLease := func(target roachpb.ReplicationTarget) {
+		if err := s.DB().AdminTransferLease(
+			ctx, startKey, target.StoreID,
+		); err != nil {
+			log.Warningf(ctx, "while transferring lease: %+v", err)
+		}
+	}
+
+	// Step 2: Repeatedly add and/or remove a replica until we reach the
+	// desired state. In an "atomic replication changes" world, this is
+	// conceptually easy: change from the old set of replicas to the new
+	// one. But there are two reasons that complicate this:
+	// 1. we can't remove the leaseholder, so if we ultimately want to do that
+	//    the lease has to be moved first. If we start out with *only* the
+	//    leaseholder, we will have to add a replica first.
+	// 2. this code is rewritten late in the cycle and it is both safer and
+	//    closer to its previous incarnation to never issue atomic changes
+	//    other than simple swaps.
+	//
+	// The loop below repeatedly calls relocateOne, which gives us either one or
+	// two ops that move the range towards the desired replication state. If
+	// it's one op, then a single add or remove is carried out (and it's only
+	// done when we can't swap instead). If it's two ops, then we're swapping
+	// (though this code doesn't concern itself with the details); and it's
+	// possible that we need to transfer the lease before we carry out the ops,
+	// determined via the leaseTarget variable.
+	//
+	// Transient errors returned from relocateOne are retried until things work
+	// out.
+	every := log.Every(time.Minute)
+	re := retry.StartWithCtx(ctx, retry.Options{MaxBackoff: 5 * time.Second})
+	for {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		ops, leaseTarget, err := s.relocateOne(ctx, &rangeDesc, targets)
+		if len(ops) == 0 {
+			// Done.
+			break
+		}
+		if fn := s.cfg.TestingKnobs.BeforeRelocateOne; fn != nil {
+			fn(ops, leaseTarget, err)
+		}
+		if leaseTarget != nil {
+			transferLease(*leaseTarget)
+		}
+
+		ctx = context.WithValue(ctx, "testing", "testing")
```

Comment:
TODO: Need to fix this hack; without it AdminChangeReplicas will run the changes one by one.

Reply:
Ok, I did something.